File: BKJN_CT/SmallLighter/2178.py
import sys
from collections import deque

sys.stdin = open("./BKJN_CT/SmallLighter/input.txt", "r") 

input = sys.stdin.readline

N, M = map(int, input().split())

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

graph = []

for _ in range(N):
    graph.append(list(map(int, input().rstrip())))

# 최소의 칸 문제 => BFS

def bfs():
    q = deque([[0,0]])
    while(q):
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < N and 0 <= ny < M and graph[nx][ny] == 1:
                q.append([nx, ny])
                graph[nx][ny] = graph[x][y] + 1

bfs()
print(graph[N - 1][M - 1])

File: BKJN_CT/SmallLighter/2178.js
const filePath =
  process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';
const input = require('fs')
  .readFileSync(filePath)
  .toString()
  .trim()
  .split('\n');

const [n, m] = input.shift().split(' ').map(Number);

const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

let graph = [];

for (let i = 0; i < n; i++) {
  graph.push(input.shift().split('').map(Number));
}

const bfs = (x, y) => {
  let q = [];
  q.push([x, y]);

  while (q.length) {
    const [x, y] = q.shift();
    for (let i = 0; i < 4; i++) {
      let nx = x + dx[i];
      let ny = y + dy[i];

      if (0 <= nx && nx < n && 0 <= ny && ny < m && graph[nx][ny] === 1) {
        q.push([nx, ny]);
        graph[nx][ny] = graph[x][y] + 1;
      }
    }
  }

  return graph[n - 1][m - 1];
};

console.log(bfs(0, 0));

let visited = new Array(n).fill(false);

const dfs = (x, y) => {
  visited[x][y] = true;
  for (let i = 0; i < 4; i++) {
    let nx = x + dx[i];
    let ny = y + dy[i];
    if (0 <= nx && nx < n && 0 <= ny && ny < m) {
      if (graph[nx][ny] === 1 && !visited[nx][ny]) {
        graph[nx][ny] = graph[x][y] + 1;
        dfs(nx, ny);
      }
    }
  }
  return graph[n - 1][m - 1];
};

console.log(dfs(0, 0));
